southworks/BotBuilder-Samples
# [FOR REVIEW] Visual Studio Extension for managing Bot Configuration

Related to #173 

## Proposed Changes

Adds a Visual Studio Extension project for managing Bot Configuration in any VS Project file.
The extensions consist of a Wizard tool which allows the users adding a  _**.bot**_ file into a specific VS Project, where they can:
 - define the **_Bot Name_**
 - declare one **_Bot Endpoints_**, specifying its _**Name**_, _**URL**_, _**AppID**_, _**AppPassword**_ (it will scale for allowing to add more than one)
 - Encrypt the _**.bot**_ file

Future features:
 - Add more than one Endpoint
 - Add more than one Service (QnA, Luis, Dispatch, etc)
 - Store the Bot Configuration information in an _**appsettings.json**_ file (for the _non sensitive_ information) along with _**secret.json**_ (using [secret manager](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-2.2&tabs=windows#secret-manager)) for the _sensitive_ information (_appPasswords_)

Here are some Screenshots of the extension Wizard:

* Here the user will specify additional information like the bot configuration filename.
![image](https://user-images.githubusercontent.com/43762887/52875642-9be3c700-3133-11e9-8845-b6fa4a002f1c.png)

* **_Endpoints_**: allows defining multiple endpoints for the bot configuration file, specifying the name, URL, appID & appPassword (these last two might be optional)
![image](https://user-images.githubusercontent.com/43762887/52875668-a900b600-3133-11e9-87f1-4f99624be16e.png)

* **_Services window_**: similar to the previous one, but allows adding any kind of service (QnA, Luis, etc)
![image](https://user-images.githubusercontent.com/43762887/52875689-b9b12c00-3133-11e9-9456-abcedfb94d8c.png)

* **_Encryption_**: allows users to encrypt the bot configuration. If they decided to do it, the encryption key will be shown _only once_.
![image](https://user-images.githubusercontent.com/43762887/52875709-c6358480-3133-11e9-95cf-54578c846886.png)

## Testing
First, go to `{your repository directory}/generators/vsix-vs-win/BotConfigurationWizard` and open the `BotFileCreator.sln` solution
Then, there are two alternatives

_Alternative A_
* Build the VSIX project
* Go to `bin/debug` folder under your project's directory and run the `BotFileCreator.vsix` file to install the extension in your Visual Studio (**NOTE**: first, ensure your Visual Studio instance is closed)
* Once the extension is installed, open any VS project file
* After the project is loaded, _right-click_ on it and select the _Create Bot File_ option in the context menu. It will open the VSIX Wizard

_Alternative B_
* Build the VSIX project
* Run the project. It will open a Visual Studio Experimental Instance
* In the Experimental instance, open any VS Project file.
* After the project is loaded, _right-click_ on it and select the _Create Bot File_ option in the context menu. It will open the VSIX Wizard

This picture shows the context menu option for starting the Extension's Wizard:
![image](https://user-images.githubusercontent.com/43762887/52876505-275e5780-3136-11e9-83c0-4d6925b19caa.png)
